src/riscv/context.rs

```diff
@@ -39,16 +40,112 @@ pub struct GeneralRegisters {
     pub t6: usize,
 }
 
+/// Floating-point registers of RISC-V.
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+pub struct FpState {
+    /// the state of the RISC-V Floating-Point Unit (FPU)
+    pub fp: [u64; 32],
+    pub fcsr: usize,
+    pub fs: FS,
+}
+
+impl Default for FpState {
+    fn default() -> Self {
+        Self {
+            fs: FS::Initial,
+            fp: [0; 32],
+            fcsr: 0,
+        }
+    }
+}
+
+#[cfg(feature = "fp-simd")]
+impl FpState {
+    /// Restores the floating-point registers from this FP state
+    #[inline]
+    pub fn restore(&self) {
+        unsafe {
+            restore_fp_registers(self);
+        }
+    }
+
+    /// Saves the current floating-point registers to this FP state
+    #[inline]
+    pub fn save(&mut self) {
+        unsafe {
+            save_fp_registers(self);
+        }
+    }
+
+    /// Clears all floating-point registers to zero
+    #[inline]
+    pub fn clear() {
+        unsafe {
+            clear_fp_registers();
+        }
+    }
+
+    /// Handles floating-point state context switching
+    ///
+    /// Saves the current task's FP state (if needed) and restores the next task's FP state
+    pub fn switch_to(&mut self, next_fp_state: &FpState) {
+        // get the real FP state of the current task
+        let current_fs = sstatus::read().fs();
+        // save the current task's FP state
+        if current_fs == FS::Dirty {
+            // we need to save the current task's FP state
+            self.save();
+            // after saving, we set the FP state to clean
+            self.fs = FS::Clean;
+        }
+        // restore the next task's FP state
+        match next_fp_state.fs {
+            FS::Clean => {
+                // the next task's FP state is clean, we should restore it
+                next_fp_state.restore();
+                // after restoring, we set the FP state
+                unsafe {
+                    sstatus::set_fs(FS::Clean);
+                }
+            }
+            FS::Initial => {
+                // restore the FP state as constant values(all 0)
+                FpState::clear();
+                // we set the FP state to initial
+                unsafe {
+                    sstatus::set_fs(FS::Initial);
+                }
+            }
+            FS::Dirty => {
+                // should not happen, since we set FS to Clean after saving
+                unreachable!("FP state of the next task should not be dirty");
+            }
+            _ => unsafe { sstatus::set_fs(FS::Off) },
```

Comment:
Write `FS::Off` here, and swap places with `FS::Dirty`, because we usually put the error branch at the last.